akka-sample-sharding-scala/src/main/scala/sample/sharding/WeatherEdges.scala

```diff
@@ -0,0 +1,145 @@
+ * In the wild, each station would run its own simple system,
+ * taking samples and posting via HTTPS.
+ */
+object WeatherEdges {
+
+  private val WeatherStations = 50
+
+  // temperature, pressure, precipitation, wind speed...
+  private val DevicesPerStation = 1
+
+  val random = new Random()
+
+  def main(args: Array[String]): Unit = {
+    val config = ConfigFactory.load("weather-station.conf")
+
+    // simple sample, using one node on local machine
+    val rootBehavior = Behaviors.setup[Nothing] { context =>
+      (0 to WeatherStations).foreach { sid =>
+        context.spawn(Station(sid, DevicesPerStation, 8081), s"station-$sid")
+      }
+      Behaviors.empty
+    }
+
+    ActorSystem[Nothing](rootBehavior, "SimulatedDevices", config)
+  }
+}
+
+object Station {
+
+  def apply(stationId: Int, devices: Int, port: Int): Behavior[WeatherHttpApi.Data] =
+    Behaviors.setup { context =>
+      context.log.info("Started station with {} devices.", devices)
+      val http = context.spawn(WeatherHttpApi(stationId, port), s"http-post-$stationId")
+
+      (0 to devices).foreach { deviceId =>
+        context.spawn(WeatherDevice(stationId, deviceId, http, 1.second, 10), s"device-$deviceId")
+      }
+
+      Behaviors.empty
+    }
+}
+
+/**
+ * Any given weather station has a unique ID, and can sample temperature, wind speed, precipitation, etc.
+ * from multiple devices.
+ *
+ * For a simple example, this does one type of reading.
+ */
+object WeatherHttpApi {
+
+  final case class Data(stationId: Int, deviceId: Int, data: List[Double])
+
+  def apply(stationId: Int, port: Int): Behavior[Data] =
+    Behaviors.setup[Data] { context =>
+      import akka.http.scaladsl.Http
+      import akka.http.scaladsl.model.{ ContentTypes, HttpEntity, HttpMethods, HttpRequest }
+
+      implicit val sys = context.system
+      import context.executionContext
+      import akka.actor.typed.scaladsl.adapter._
+      val http = Http(context.system.toClassic)
+
+      // Run and completely consume a single akka http request
+      def runRequest(req: HttpRequest) =
+        http.singleRequest(req).flatMap { _.entity.dataBytes.runReduce(_ ++ _) }
+
+      Behaviors.receiveMessage[Data] {
+        case Data(sid, did, data) =>
+          // TODO marshalling vs scary hand-coded!
+          val values = data.mkString(",")
+          val json = s"""{"stationId":$sid,"deviceId":$did,"data":[$values]}"""
+          context.log.info("Read: {}", json)
+
+          // See https://doc.akka.io/docs/akka-http/current/common/index.html
+          // for marshalling, unmarshalling, json support in the wild.
+          val httpEntity = HttpEntity(ContentTypes.`application/json`, json)
+          runRequest(
+            HttpRequest(method = HttpMethods.POST, uri = s"http://localhost:$port/temperatures", entity = httpEntity))
+            .onComplete {
+              case Success(response) =>
+                context.log.info(response.utf8String)
```

Comment:
We had some comments about context.log from other thread elsewhere.

Reply:
Yes, done.